server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -213,20 +270,80 @@ remoteSchemaInterface schemaDoc defn@(G.InterfaceTypeDefinition description name
+      -> (SelectionSet NoFragments G.Name)
+    constructInterfaceSelectionSet objNameAndSelSets =
+      let selSetFieldsWithoutFragments' =
+            flip (fmap . fmap . fmap) objNameAndSelSets $
+              \case
+                -- We're omitting a selection field with fragments
+                -- here because the `SelectionSet` we recieve is
+                -- a "flattened" selection set without any fragments
+                -- (atleast at the top level of the selection set).
+                -- This is needed to extract the common interface
+                -- fields from the various objects that have been
+                -- queried.
+                G.SelectionField fld -> Just fld
+                _                    -> Nothing
+          selSetFieldsWithoutFragments = (fmap . fmap) catMaybes selSetFieldsWithoutFragments'
+
+          -- common interface fields that exist in every
+          -- selection set provided
+          -- #1 of Note [Querying remote schema Interfaces]
+          commonInterfaceFields =
+            Map.elems $
+            Map.mapMaybe allTheSame $
+            Map.groupOn G._fName $
+            concatMap (\(_, flds) ->
+                         filter ((`elem` interfaceFieldNames) . G._fName) flds)
+                         $ selSetFieldsWithoutFragments
+
+          interfaceFieldNames = map G._fldName fields
+
+          allTheSame [] = Nothing
+          allTheSame (x:xs) = bool Nothing (Just x) $ all (== x) xs
+
+          -- #2 of Note [Querying remote schema interface fields]
+          nonCommonInterfaceFields =
+            catMaybes $ flip map selSetFieldsWithoutFragments $ \(objName, flds) ->
+              let nonCommonFields = filter (not . flip elem commonInterfaceFields) flds
+              in mkObjInlineFragment (objName, map G.SelectionField nonCommonFields)
+
+          -- helper function for #4 of Note [Querying remote schema interface fields]
+          mkObjInlineFragment (_, []) = Nothing
+          mkObjInlineFragment (objName, selSet) =
+            Just $ G.SelectionInlineFragment $
+              G.InlineFragment (Just objName) mempty selSet
+
+      -- #5 of Note [Querying remote schema interface fields]
+      in (map G.SelectionField commonInterfaceFields) <> nonCommonInterfaceFields
+
 -- | 'remoteSchemaUnion' returns a output parser for a given 'UnionTypeDefinition'.
 remoteSchemaUnion
   :: forall n m
   . (MonadSchema n m, MonadError QErr m)
   => SchemaIntrospection
   -> G.UnionTypeDefinition
-  -> m (Parser 'Output n ())
+  -> m (Parser 'Output n (SelectionSet NoFragments G.Name))
 remoteSchemaUnion schemaDoc defn@(G.UnionTypeDefinition description name _directives objectNames) =
   P.memoizeOn 'remoteSchemaObject defn do
-  objDefs <- traverse getObject objectNames
-  objs :: [Parser 'Output n ()] <- traverse (remoteSchemaObject schemaDoc) objDefs
+  objs :: [Parser 'Output n (Name, (SelectionSet NoFragments G.Name))] <-
```

Comment:
```suggestion
  objs :: [Parser 'Output n (Name, SelectionSet NoFragments G.Name)] <-
```

Reply:
Removing the type annotation from here, because moved the logic to `getObject` function